kitsune/wiki/forms.py

```diff
@@ -158,20 +153,34 @@ def clean_slug(self):
         return slug
 
     def clean(self):
-        c = super(DocumentForm, self).clean()
-        locale = c.get("locale")
+        cdata = super(DocumentForm, self).clean()
+        locale = cdata.get("locale")
 
         # Products are required for en-US
-        products = c.get("products")
-        if locale == settings.WIKI_DEFAULT_LANGUAGE and (not products or len(products) < 1):
+        product_ids = cdata.get("products")
+        products = Product.objects.filter(id__in=[int(product_id) for product_id in product_ids])
+        if locale == settings.WIKI_DEFAULT_LANGUAGE and (not product_ids or len(product_ids) < 1):
             raise forms.ValidationError(PRODUCT_REQUIRED)
 
         # Topics are required for en-US
-        topics = c.get("topics")
-        if locale == settings.WIKI_DEFAULT_LANGUAGE and (not topics or len(topics) < 1):
+        topic_ids = cdata.get("topics")
+        topics = Topic.objects.filter(id__in=[int(topic_id) for topic_id in topic_ids])
+        if locale == settings.WIKI_DEFAULT_LANGUAGE and (not topic_ids or len(topic_ids) < 1):
             raise forms.ValidationError(TOPIC_REQUIRED)
 
-        return c
+        invalid_topics = []
+        for topic in topics:
+            associated_products = topic.products.all()
+            if not set(products).intersection(associated_products):
```

Comment:
I think `associated_products` has to be a set here?

Reply:
Good catch. I removed the `set()` completely in order to use Django's `intersection()` method since we are operating on QuerySets.